File: mergeintervals.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        intervals = sorted(intervals, key=lambda x: x.start)
        res = []
        while len(intervals):
            range = intervals.pop(0)
            for i in res:
                if range.start <= i.end:
                    i.end = range.end if range.end > i.end else i.end
                    break;
            else:
                res.append(range)
        return res

File: mergeintervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
        sort(intervals.begin(), intervals.end(), [](const Interval& a, const Interval& b) {return a.start < b.start;});
        vector<Interval> result;
        for (int i = 0; i < intervals.size(); i++) {
            if (result.empty()) {
                result.push_back(intervals[i]);
                continue;
            }
            if (intervals[i].start > result.back().end) result.push_back(intervals[i]);
            else{
                if (intervals[i].end > result.back().end) result.back().end = intervals[i].end;
            }
        }
        return result;
    }
};
